### PR TITLE
perf(oxc_semantic): `get_symbol_id_from_span` with binary search

### PR DESCRIPTION
**Pull request by IWANABETHATGUY**

### CODE REVIEW

File: crates/oxc_semantic/src/symbol.rs [hunk 1]
```diff
@@ -62,10 +62,13 @@ impl SymbolTable {
...
 
+    // Symbol are created in dfs order, so we make tow assumptions:
+    // 1. The start of [SymbolTable.spans] is strictly monotonically increasing.
+    // 2. There are no two symbols with the same span.
+    // So we could use binary search to find the symbol id.
     pub fn get_symbol_id_from_span(&self, span: Span) -> Option<SymbolId> {
-        self.spans
-            .iter_enumerated()
-            .find_map(|(symbol, &inner_span)| if inner_span == span { Some(symbol) } else { None })
+        let index = self.spans.raw.binary_search_by(|a| a.start.cmp(&span.start)).ok()?;
```
Comment (camc314) [inline crates/oxc_semantic/src/symbol.rs:70]: i wonder whether a debug assert would be a good idea here to validate the assumptions

File: crates/oxc_semantic/src/symbol.rs [hunk 2]
```diff
@@ -62,10 +62,13 @@ impl SymbolTable {
         self.spans.iter_enumerated().map(|(symbol_id, _)| symbol_id)
     }
 
+    // Symbol are created in dfs order, so we make tow assumptions:
```
Comment (camc314) [inline crates/oxc_semantic/src/symbol.rs:65]:
```suggestion
    // Symbol are created in dfs order, so we make two assumptions:

File: crates/oxc_semantic/src/symbol.rs [hunk 3]
```diff
@@ -62,10 +62,13 @@ impl SymbolTable {
         self.spans.iter_enumerated().map(|(symbol_id, _)| symbol_id)
     }
 
+    // Symbol are created in dfs order, so we make two assumptions:
```
Comment (Boshen) [inline crates/oxc_semantic/src/symbol.rs:65]: Let's not make this assumption.

This API is added for the import plugin, are you using this anywhere else?

Comment (overlookmotel) [reply crates/oxc_semantic/src/symbol.rs:65]: Yes. That assumption will not hold if the AST has been through transformer. Right now I believe Rolldown re-runs semantic after transformer, but we are working to keep `Semantic` in sync with AST in transformer, to make that unnecessary (which will be a decent speed-up).

**Pull request closed without merging**